File: generated_code/GeminiPro/Python/042_SplayTree.py
class SplayTree:
    """
    A self-contained Splay Tree class implementing a dictionary-like set for integers.

    This implementation includes insert, delete, and search operations. The search
    operation (and consequently insert and delete) performs a splay, moving the
    accessed node (or its parent/successor) to the root of the tree to optimize

    future accesses.
    """

    class _Node:
        """Internal class to represent a node in the Splay Tree."""
        def __init__(self, key, parent=None, left=None, right=None):
            self.key = key
            self.parent = parent
            self.left = left
            self.right = right

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left is not None:
            y.left.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right is not None:
            y.right.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Splays the node x to the root of the tree.
        This operation involves a sequence of rotations (Zig, Zig-Zig, Zig-Zag)
        to move the node up the tree.
        """
        while x.parent is not None:
            p = x.parent
            g = p.parent
            if g is None:  # Zig step
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-Zig step
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-Zig step
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-Zag step
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-Zag step (x == p.left and p == g.right)
                self._right_rotate(p)
                self._left_rotate(g)

    def search(self, key):
        """
        Searches for a key in the tree and splays the accessed node.

        If the key is found, the node containing the key is splayed to the root.
        If the key is not found, the last non-null node visited during the
        search is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node is not None:
            last_node = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                # Key found, splay the node and return True
                self._splay(node)
                return True

        # Key not found, splay the last visited node (if any)
        if last_node is not None:
            self._splay(last_node)
        
        # After splaying, the root might be the key, but search failed before.
        # The return value should reflect the initial search result.
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted as a new node and then splayed to
        the root. This class implements a set, so duplicates are not stored.

        Args:
            key: The integer key to insert.
        """
        node = self.root
        parent = None
        while node is not None:
            parent = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                # Key already exists, splay it to the root
                self._splay(node)
                return

        # Insert the new key
        new_node = self._Node(key, parent)
        if parent is None:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        The tree is first splayed on the key. If the key is found, it becomes
        the root. The root is then removed, and its left and right subtrees
        are joined together.

        Args:
            key: The integer key to delete.
        """
        # Splay the tree on the key. This brings the node (or its neighbor) to the root.
        # We perform the search-and-splay manually here to avoid an extra check.
        node = self.root
        last_node = None
        while node is not None:
            last_node = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                break # Found the node
        
        if last_node is not None:
            self._splay(last_node)
        
        # After splaying, if the root's key is not the target key, it wasn't in the tree.
        if self.root is None or self.root.key != key:
            return

        # The node to delete is now at the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if left_subtree is None:
            self.root = right_subtree
            if self.root is not None:
                self.root.parent = None
        elif right_subtree is None:
            self.root = left_subtree
            if self.root is not None:
                self.root.parent = None
        else:
            # Disconnect both subtrees from the old root
            left_subtree.parent = None
            right_subtree.parent = None

            # Find the maximum element in the left subtree
            max_in_left = left_subtree
            while max_in_left.right is not None:
                max_in_left = max_in_left.right

            # Splay this maximum element to the root of the (now detached) left subtree.
            # We can temporarily set self.root to perform the splay in the correct context.
            self.root = left_subtree
            self._splay(max_in_left)

            # The new root is the max element from the left subtree.
            # Attach the original right subtree as its right child.
            self.root.right = right_subtree
            right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/042_SplayTree.java
/**
 * Implements a Splay Tree, a self-balancing binary search tree, for storing a set of integers.
 * Splay trees have the property that recently accessed elements are quick to access again.
 * All major operations (insert, delete, search) perform a "splay" operation, which moves
 * the accessed element (or the last element on the search path) to the root of the tree.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splay operation. Moves the node with the given key (or the last accessed
     * node on the search path) to the root of the subtree.
     * This is a top-down splay implementation.
     * @param root The root of the subtree to splay.
     * @param key The key to splay around.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node root, int key) {
        if (root == null) {
            return null;
        }

        // A dummy node to simplify linking of left and right subtrees.
        Node header = new Node(0);
        Node leftTreeMax = header;
        Node rightTreeMin = header;
        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) {
                    break;
                }
                // Zig-Zig (left-left)
                if (key < current.left.key) {
                    current = rightRotate(current);
                    if (current.left == null) {
                        break;
                    }
                }
                // Link to the right tree
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    break;
                }
                // Zig-Zig (right-right)
                if (key > current.right.key) {
                    current = leftRotate(current);
                    if (current.right == null) {
                        break;
                    }
                }
                // Link to the left tree
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else { // key == current.key
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = header.right;
        current.right = header.left;

        return current;
    }

    /**
     * Searches for a key in the tree. If found, the node is splayed to the root.
     * If not found, the last accessed node is splayed to the root.
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree. If the key already exists, the tree remains unchanged.
     * The newly inserted node becomes the new root of the tree.
     * @param key The key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. The closest node will be the root.
        root = splay(root, key);

        // If the key is already present, do nothing.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree. If the key is not found, the tree remains unchanged.
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree around the key. If it exists, it will be the root.
        root = splay(root, key);

        // If the key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        // The node to be deleted is now the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Splay the largest element in the left subtree to its root.
            // This is done by splaying for the key we are deleting, which is guaranteed
            // to be larger than any element in the left subtree.
            Node newRoot = splay(leftSubtree, key);
            newRoot.right = rightSubtree;
            root = newRoot;
        }
    }
}